### PR TITLE
bity.com crypto off-ramping integration

### LINKED ISSUE #107: Integrate bity FIAT off-ramp
## Scope

- [Here](https://github.com/austintgriffith/burner-wallet/issues/177) is the original idea for integrating bity
- Bity allows a user to off-ramp ETH <=> 5000 CHF annually. It only requires authentication through a mobile phone number
- Bity only works in the EU at this point. Make this sufficiently clear to the user.
- Use bity's API to off-ramp a user's eth as outlined in their documentation: https://doc.bity.com/backend/v2.html#place-an-order
- Consider storing some of the user's credentials (e.g. phone number) in the wallet's local storage for even faster off-ramping.
- Discuss whether or not it is safe to store the user's bank account of second off-ramping too. Personally, I think it's not secure though.
- Bonus: Reach out to bity and ask if they're interested in integrating with our main net for even faster off-ramping
- Make the user aware of the annual off-ramping limit of 5000 CHF by querying the past orders of the user and summing up their withdrawals to show it in the UI

## Deliverables

- A dialogue in the Exchange view of the burner wallet that lets the user off-ramp crypto into fiat very quickly

## Gain for the project

- Allows users to exchange their crypto for FIAT without much friction
- Proves to newly onboarded users that pDAI/ETH is in fact valuable

## Roles
bounty gardener: @TimDaub / 10%
bounty worker: name / 75%
bounty reviewer: name / 15%

### PR DESCRIPTION
**Pull request by TimDaub**
Fixes #107

Checkout
Screen Shot 2019-06-27 at 18 09 28

Receipt
screen_shot_2019-06-27_at_16 45 44

Order Status
Screen Shot 2019-06-27 at 18 08 11

Todos:

- [x] Change all bootstrap usage to rimble-ui
- [x] Make dialogue show button and not prompt when opening Exchange.js
- [x] Make user confirm transaction to bity
- [x] Do proper validation of inputs to avoid user errors
- [x] Remove unnecessarily included contract files again
- [x] Make bity integration i18n ready
- [x] Remove Exchange.js bity integration left-overs
- [x] Cursor on IBAN form resets when changing text in the middle
- [x] Rename `Cashout.js` to e.g. `Bity.js`
- [x] Get minimum possible order amount through API
- [x] Store bity order in local storage and allow user to check it's history and status
  - [x] Get correct block timestamp for bity order (it's mainnet not Plasma chain)
- [x] Resolve all TODOs in the code
- [x] Propagate all errors to the user
- [x] Custom `<Loader>` time? Or Serviceworker that transfers in the background?
- [x] Rebase against current master
- [x] Give annual amount indicator or error (maybe by quering list orders)
- [x] It's not clear whether test bity.com order went through (e.g. bank account name is missing) (it went through)

## Problems

I've created issues for problems that I deemed were not solveable in the scope of this PR:

- "Display bity.com annual cashout limit to user": https://github.com/leapdao/burner-wallet/issues/171
- "Use ethgasstation to predict wait times of Ethereum Mainnet transactions correctly": https://github.com/leapdao/burner-wallet/issues/170
- "Allow user to pick their local currency" https://github.com/leapdao/burner-wallet/issues/169
- "`<RecentTransaction />` time is based on side chain block number": https://github.com/leapdao/burner-wallet/issues/172
- "Use react-i18next instead of i18n": https://github.com/leapdao/burner-wallet/issues/173
- "bity.com cashout functionality makes the user wait too long": https://github.com/leapdao/burner-wallet/issues/183
- "Cashout entire balance through bity.com": #184 184

### CODE REVIEW

File: src/components/Bity.js [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+  "SM",
+  "ES",
+  "SK",
+  "SI",
+  "SE",
+  "CH",
+  "GB"
+];
+
+let MIN_AMOUNT_DOLLARS = 15;
```
Comment (TimDaub) [inline src/components/Bity.js]: Please note that this value (15) gets overwritten when `componentDidMount` is called.

File: src/components/Bity.js [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+let prevFormattedIBAN;
+
+class Bity extends Component {
+  constructor(props) {
+    super(props);
+
+    const pk = localStorage.getItem("metaPrivateKey");
```
Comment (TimDaub) [inline src/components/Bity.js]: Really unhappy with this. But it's currently the way burner-modules are supposed to be written :/

Comment (iamonuwa) [reply src/components/Bity.js]: Can we have a quick call about this?

File: src/components/Bity.js [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+      return newFields;
+    };
+  }
+
+  render() {
+    const { fields } = this.state;
+    return (
+      <div>
+        <Box mb={4}>
+          {/* TODO: How to put this into i18n without creating a mess?*/}
```
Comment (TimDaub) [inline src/components/Bity.js]: Still unclear how this can be done with `i18n`. I'd propose to just ignore it for now...

Comment (TimDaub) [reply src/components/Bity.js]: @iamonuwa I did a bit more research and came up with the following issue: https://github.com/leapdao/burner-wallet/issues/173

File: src/components/Bity.js (hunk 2 above)
Comment (TimDaub) [reply src/components/Bity.js]: @iamonuwa and I had a call. I connected him with @dmihal who's leading the burner-core integration.

File: src/components/Bity.js (hunk 3 above)
Comment (iamonuwa) [reply src/components/Bity.js]: i18next is what was used in the initial implementation. That's what the internationalization is running on

Comment (iamonuwa) [reply src/components/Bity.js]: I'll take a look at this and see if there is a better solution to it

**Comment by iamonuwa** [comment]
> Testing off-ramping functionality on mainnet with metamask:
> 
> * Try to exit ETH to Fiat ($25)
> 
> At the stage or pasting the IBAN there's a warning in the console:
> 
> ```
> index.js:1437 Warning: A component is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
>     in input (created by Context.Consumer)
>     in StyledComponent (created by Input)
>     in Input (at Bity.js:378)...
> ```
> 
> Then after performing the exchange and metamask confirmation, got this:
> 
> ```
> Uncaught (in promise) Error: Returned error: Error: Error: [ethjs-rpc] rpc error with payload {"id":5958094193946,"jsonrpc":"2.0","params":["0xf869328477359400809442e17e25f885a0be481867bdc8aab458c442996988012baa5ac39f67d68026a0d9800e725290a7aeae77538a8501de877a8eb4c4c075b6502ba597e30cc6a5bba005a2c1138ebb93fdbc331ee5c73290f83ad5fd9d80a0353ec3a9d7ba12da214c"],"method":"eth_sendRawTransaction"} {"code":-32000,"message":"intrinsic gas too low"}
> ```
> 
> Burner's interface stalled with this picture:
> Screen Shot 2019-07-04 at 12 17 47
> 
> and there's an error in metamask as well:
> Screen Shot 2019-07-04 at 12 18 16
> 
> Money not lost :)
> However expected the transaction to be performed...
> 
> Later observed this in the console (maybe related):
> 
> ```
> Error: Returned values aren't valid, did it run Out of Gas?
>     at ABICoder.push../node_modules/web3-eth-abi/src/index.js.ABICoder.decodeParameters (index.js:227)
> ```

Which network are you testing with?

**Comment by articobandurini** [comment]
mainnet, as i noted in the first sentence :)

**Comment by iamonuwa** [comment]
Okay,

**Pull request merged**